File: src/lib/wireguard/record.ts
import { Output, Resource } from '@pulumi/pulumi';

import { dnsConfig } from '../configuration';
import { createRecord } from '../google/dns/record';

/**
 * Creates the DNS records.
 *
 * @param {Output<string>} ipv4Address the IPv4 address of the WireGuard instance
 * @param {Output<string>} ipv6Address the IPv6 address of the WireGuard instance
 * @returns {Output<Resource[]>} the DNS records
 */
export const createWireguardDNSRecords = (
  ipv4Address: Output<string>,
  ipv6Address: Output<string>,
): Resource[] => [
  createRecord(
    dnsConfig.entries.wireguard.domain,
    dnsConfig.entries.wireguard.zoneId,
    'A',
    [ipv4Address],
    {
      project: dnsConfig.project,
    },
  ),

  createRecord(
    dnsConfig.entries.wireguard.domain,
    dnsConfig.entries.wireguard.zoneId,
    'AAAA',
    [ipv6Address],
    {
      project: dnsConfig.project,
    },
  ),
];

File: pkg/lib/wireguard/record.go
package wireguard

import (
	"github.com/muhlba91/pulumi-shared-library/pkg/lib/google/dns/record"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"

	dnsConf "github.com/muhlba91/muehlbachler-core-infrastructure/pkg/model/config/dns"
)

// CreateDNSRecords creates DNS records for Wireguard based on the provided DNS configuration.
// ctx: The Pulumi context for resource creation.
// dnsConfig: The DNS configuration containing domain and record details.
// ipv4: The public IPv4 address to create DNS records for.
// ipv6: The public IPv6 address to create DNS records for.
func CreateDNSRecords(
	ctx *pulumi.Context,
	dnsConfig *dnsConf.Config,
	ipv4 pulumi.StringOutput,
	ipv6 pulumi.StringOutput,
) []pulumi.Resource {
	dnsEntry := dnsConfig.Entries["wireguard"]

	v4, v4Err := record.Create(ctx, &record.CreateOptions{
		Domain:     *dnsEntry.Domain,
		ZoneID:     pulumi.String(*dnsEntry.ZoneID),
		RecordType: "A",
		Records:    pulumi.StringArray([]pulumi.StringInput{ipv4}),
		Project:    dnsConfig.Project,
	})
	if v4Err != nil {
		return nil
	}

	v6, v6Err := record.Create(ctx, &record.CreateOptions{
		Domain:     *dnsEntry.Domain,
		ZoneID:     pulumi.String(*dnsEntry.ZoneID),
		RecordType: "AAAA",
		Records:    pulumi.StringArray([]pulumi.StringInput{ipv6}),
		Project:    dnsConfig.Project,
	})
	if v6Err != nil {
		return nil
	}

	return []pulumi.Resource{v4, v6}
}
